src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -4293,6 +4293,66 @@ TEST_CASE("AdaptiveMeanPoolingTestCase", "[ANNLayerTest]")
   REQUIRE(arma::accu(delta) == 2.25);
 }
 
+/**
+ * Simple Separable Convolution layer test.
+ */
+TEST_CASE("SimpleSeparableConvolutionLayerTest", "[ANNLayerTest]")
+{
+  arma::mat input(224 * 224 * 3, 1), output;
+  SeparableConvolution<> module1(3, 21, 3, 3, 1, 1, 0, 0, 224, 224, 3);
+
+  // Test the forward function.
+  input.fill(2);
+  module1.Parameters().fill(1);
+  module1.Reset();
+  module1.Forward(input, output);
+
+  // Comparison value taken from PyTorch conv layer with the same config
+  REQUIRE(arma::accu(output) == 19664316);
+}
+
+/**
+ * Separable Convolution layer numerical gradient test.
+ */
+TEST_CASE("GradientSeparableConvolutionLayerTest", "[ANNLayerTest]")
+{
+  struct GradientFunction
+  {
+    GradientFunction() :
+        input(arma::randu(20 * 20 * 2, 1)),
+        target(arma::mat("0"))
+    {
+      model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>();
+      model->Predictors() = input;
+      model->Responses() = target;
+      model->Add<SeparableConvolution<>>(2, 4, 3, 3, 1, 1, 0, 0, 20, 20, 2);
+      model->Add<LogSoftMax<>>();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      double error = model->Evaluate(model->Parameters(), 0, 1);
+      model->Gradient(model->Parameters(), 0, gradient, 1);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+    arma::mat input, target;
+  } function;
+
+  // I understand that this is quite a bit more but wasn't really sure
+  // on how to write the test in the first place so if something is wrong
+  // with that do let me know.
```

Comment:
@zoq, @kartikdutt18 I added the test and also added a comment on the tolerance of `0.4`.

Reply:
This is quite large, do you have the output of `CheckGradient(function)`?

Reply:
It was something around `0.39xx`

Reply:
The thing is I am not sure if the test case is written properly, so if you could help me rectify that it would be more helpful.

Reply:
The test implementation is correct.